src/audio/drc/drc.c

```diff
@@ -0,0 +1,606 @@
+					     int channels)
+{
+	int i;
+
+	/* Allocate pre-delay (lookahead) buffers */
+	for (i = 0; i < channels; ++i) {
+		state->pre_delay_buffers[i] =
+			rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM,
+				sample_bytes * DRC_MAX_PRE_DELAY_FRAMES);
+		if (!state->pre_delay_buffers[i])
+			return -ENOMEM;
+		memset(state->pre_delay_buffers[i], 0, sample_bytes * DRC_MAX_PRE_DELAY_FRAMES);
+	}
+
+	return 0;
+}
+
+static inline int drc_set_pre_delay_time(struct drc_state *state,
+					 int32_t pre_delay_time,
+					 int32_t rate)
+{
+	int32_t pre_delay_frames;
+
+	/* Re-configure look-ahead section pre-delay if delay time has changed. */
+	pre_delay_frames = Q_MULTSR_32X32((int64_t)pre_delay_time, rate, 30, 0, 0);
+	if (pre_delay_frames < 0)
+		return -EINVAL;
+	pre_delay_frames = min(pre_delay_frames, DRC_MAX_PRE_DELAY_FRAMES - 1);
+
+	/* Make pre_delay_frames multiplies of DIVISION_FRAMES. This way we
+	 * won't split a division of samples into two blocks of memory, so it is
+	 * easier to process. This may make the actual delay time slightly less
+	 * than the specified value, but the difference is less than 1ms. */
+	pre_delay_frames &= ~DRC_DIVISION_FRAMES_MASK;
+
+	/* We need at least one division buffer, so the incoming data won't
+	 * overwrite the output data */
+	pre_delay_frames = max(pre_delay_frames, DRC_DIVISION_FRAMES);
+
+	if (state->last_pre_delay_frames != pre_delay_frames) {
+		state->last_pre_delay_frames = pre_delay_frames;
+		state->pre_delay_read_index = 0;
+		state->pre_delay_write_index = pre_delay_frames;
+	}
+	return 0;
+}
+
+static int drc_setup(struct comp_data *cd, uint16_t channels, uint32_t rate)
+{
+	int ret = 0;
+
+	/* Reset any previous state */
+	drc_reset_state(&cd->state);
+
+	/* Allocate pre-delay buffers */
+	uint32_t sample_bytes = get_sample_bytes(cd->source_format);
+	ret = drc_init_pre_delay_buffers(&cd->state, (size_t)sample_bytes, (int)channels);
+	if (ret < 0)
+		return ret;
+
+	/* Set pre-dely time */
+	return drc_set_pre_delay_time(&cd->state, cd->config->params.pre_delay_time, rate);
+}
+
+/*
+ * End of DRC setup code. Next the standard component methods.
+ */
+
+static struct comp_dev *drc_new(const struct comp_driver *drv,
+				struct sof_ipc_comp *comp)
+{
+	struct comp_dev *dev;
+	struct comp_data *cd;
+	struct sof_ipc_comp_process *drc;
+	struct sof_ipc_comp_process *ipc_drc =
+		(struct sof_ipc_comp_process *)comp;
+	size_t bs = ipc_drc->size;
+	int ret;
+
+	comp_cl_info(&comp_drc, "drc_new()");
+
+	/* Check first before proceeding with dev and cd that coefficients
+	 * blob size is sane.
+	 */
+	if (bs > SOF_DRC_MAX_SIZE) {
+		comp_cl_err(&comp_drc, "drc_new(), coefficients blob size %u exceeds maximum",
+			    bs);
+		return NULL;
+	}
+
+	dev = comp_alloc(drv, COMP_SIZE(struct sof_ipc_comp_process));
+	if (!dev)
+		return NULL;
+
+	drc = COMP_GET_IPC(dev, sof_ipc_comp_process);
+	ret = memcpy_s(drc, sizeof(*drc), ipc_drc,
+		       sizeof(struct sof_ipc_comp_process));
```

Comment:
i am not sure if i am missing something but I dont see where drc get used/stored? Are we leaking here?

Reply:
The code is performing memory copy from ipc_drc (obtained from drc_new() input) to dev->comp (drc = dev->comp)
dev->comp will be used in future process.

Reply:
ah yes it is much clearer with the update, thanks